packages/react-dom/src/shared/ensureCorrectIsomorphicReactVersion.js

```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import reactDOMPackageVersion from 'shared/ReactVersion';
+import * as IsomorphicReactPackage from 'react';
+
+export function ensureCorrectIsomorphicReactVersion() {
+  const isomorphicReactPackageVersion = IsomorphicReactPackage.version;
+  if (isomorphicReactPackageVersion !== reactDOMPackageVersion) {
+    // TODO: Do we have a canoncical documentation page for this?
+    throw new Error(
+      'Incompatible React versions: The "react" and "react-dom" packages must ' +
```

Comment:
In a perfect world we'd also detect environment mismatches e.g. importing `react-dom` **with** the `react-server` condition while `react` is being imported **without** the `react-server` condition. Maybe we encode the environment in the version string e.g. `19.0.0-abc-123+react-server`? Though that wouldn't tell you explcitly that the environment mismatched not the version.

Reply:
We could add another internal state field, but I'm less concerned about that one since it's only the tip of the iceberg of what you have to do to configure a Server Components set-up correctly. For a similar reason I didn't add a check to React Native because nobody really imports the React Native renderer directly; it's configured by some sort of framework.

Reply:
Like, RN installs the renderer, but the template still allows using your own version of `react`, and we should catch that. This will start erroring in the OSS build on land, but that needs to get fixed before the next RN npm release.